cronjob_scripts/stats.py

```diff
@@ -12,40 +17,71 @@
 _influxdb_client = None
 
 
-def get_stats(period: str, arch: str | None):
+@lru_cache
+def get_stats(period: str) -> DataFrame:
+    """
+    To reduce the number of round-trips to the InfluxDB server, we do a single query and cache the
+    results of this function.
+
+    In theory it would be better if we could use a dataloader-like pattern to batch up requests and
+    ask for exactly the architectures we want, but that would require us to make the whole script
+    async, and we know that cronjob will eventually ask for all architectures anyway.
+    """
     global _influxdb_client
     if _influxdb_client is None:
         _influxdb_client = InfluxDBClient3(
             host=HOST, token=TOKEN, org=ORG, database=DATABASE
         )
 
+    # Old clients (going via the old stats server) don't report status, so we will end up triggering
+    # re-checks even for things successfully installed from tarball. Hopefully these will gradually
+    # get phased out and then this method will only return crate versions which ended up installing
+    # from source.
+    #
+    # FIXME: seems like there are some entries of the form:
+    # {"agent": "binstalk-fetchers/0.10.0", "status": "start", ... }
+    # Apparently binswap-github is using an old version of binstalk. See:
+    # https://github.com/cargo-bins/cargo-quickinstall/pull/300/files#r1778063083
     query = """
-        SELECT DISTINCT crate
+        SELECT DISTINCT crate, target, version, status
```

Comment:
I feel like `DISTINCT ON` is what we actually want, but not sure if it is supported

```suggestion
        SELECT DISTINCT ON(crate, target) crate, target, version, status
```

Reply:
I think status and version will both be very low cardinality (when grouped by (crate, target) for the last day). 

Generally we always ask for the latest version, so there will be either 1 or 2 different values of `version` per crate, depending on whether there was a release today.

If a crate version is unbuilt then we will receive some status=null spam from old clients and status=not-found from new clients. I think this will result in no more than twice the rows back from influxdb compared to your method, so I don't think it's worth optimising unless influxdb starts rate limiting us (which I am not expecting to happen if we make this query once per hour). We need to keep both statuses separate so that we can construct the 'not-found' high priority short list.

Reply:
Doesn't `DISTINCT crate` remove all rows that have the same crate name?

I.e. if two rows have different target names but the same crate name, only one of them will be kept

Reply:
in  postgres no: https://www.postgresql.org/docs/current/sql-select.html#SQL-DISTINCT

        SELECT DISTINCT crate, target, version, status

would be equivalent to

        SELECT DISTINCT on(crate, target, version, status) crate, target, version, status

Reply:
Thanks, if this is the case then I'm ok with this SQL code